tests/test_webext.py

```diff
@@ -0,0 +1,171 @@
+# Copyright Mozilla Foundation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from importlib.resources import files
+from textwrap import dedent
+from unittest import TestCase
+
+from moz.l10n.message import Declaration, Expression, PatternMessage, VariableRef
+from moz.l10n.resource import Entry, Resource, Section
+from moz.l10n.webext import webext_parse, webext_serialize
+
+# Show full diff in self.assertEqual. https://stackoverflow.com/a/61345284
+# __import__("sys").modules["unittest.util"]._MAX_LENGTH = 999999999
+
+source = files("tests.data").joinpath("messages.json").read_bytes()
+
+
+class TestWebext(TestCase):
+    def test_parse(self):
+        res = webext_parse(source)
+        self.assertEqual(
+            res,
+            Resource(
+                [
+                    Section(
+                        [],
+                        [
+                            Entry(
+                                ["SourceString"],
+                                PatternMessage(["Translated String"]),
+                                comment="Sample comment",
+                            ),
+                            Entry(
+                                ["MultipleComments"],
+                                PatternMessage(["Translated Multiple Comments"]),
+                                comment="Second comment",
+                            ),
+                            Entry(
+                                ["NoCommentsorSources"],
+                                PatternMessage(["Translated No Comments or Sources"]),
+                            ),
+                            Entry(
+                                ["placeholders"],
+                                PatternMessage(
+                                    [
+                                        "Hello$$ ",
```

Comment:
It puzzles me why is this not `"Hello$$$ "`.

Reply:
From [here](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/i18n/Locale-Specific_Message_reference#message):
> Any number of consecutive dollar signs appearing in strings are replaced by the same number of dollar signs minus one. So, $$ > $, $$$ > $$, etc.